### PR TITLE
Reinitialize FormData stream on 307 redirects

### PR DESCRIPTION
**Pull request by codenirvana**
## PR Checklist:
- [x] I have run `npm test` locally and all tests are passing.
- [x] I have added/updated tests for any new behavior.
       
- [x] If this is a significant change, an issue has already been created where the problem / solution was discussed: https://github.com/postmanlabs/postman-app-support/issues/4077
       

## PR Description
Reinitialize FormData stream for 307 redirects

### CODE REVIEW

File: tests/test-form-data-redirect.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        }
+      }
+    }
+
+    var reqOptions = {
+      url: url + options.url,
+      formData: multipartFormData,
+      followAllRedirects: true
+    }
+    request.post(reqOptions, function (err, res, body) {
```
Comment (kunagpal) [inline tests/test-form-data-redirect.js]: `reqOptions` can be replaced with the literal object.

File: lib/redirect.js [hunk 2]
```diff
@@ -150,6 +152,66 @@ Redirect.prototype.onResponse = function (response) {
         request.removeHeader('authorization')
       }
     }
+  } else if (request.formData &&
```
Comment (kunagpal) [inline lib/redirect.js]: Add:
1. TODOs to expose these errors and edge cases as events.
2. Boundary comments explaining that this is a Postman request specific change.
3. Updates to the fork differences list with the same (if not done already)

File: lib/redirect.js [hunk 3]
```diff
@@ -150,6 +152,66 @@ Redirect.prototype.onResponse = function (response) {
...
+      // make sure _form is released and there's no pending _streams left
+      // which will be the case for 401 redirects. so, reuse _form on redirect
```
[the remark this answers is not in the document]
Comment (codenirvana) [reply lib/redirect.js]:
```
If the request included authentication credentials, then the 401
response indicates that authorization has been refused for those
credentials.  The user agent MAY repeat the request with a new or
replaced Authorization header field
```
https://tools.ietf.org/html/rfc7235#section-3.1

Here, configured by `sendImmediately` option.

File: lib/redirect.js [hunk 4]
```diff
@@ -150,6 +152,66 @@ Redirect.prototype.onResponse = function (response) {
...
+      if (!Array.isArray(formData[key])) {
+        formData[key] = [formData[key]]
+      }
+
+      formData[key].push(newValue)
+    }
+    for (var formKey in request.formData) {
+      if (request.formData.hasOwnProperty(formKey)) {
+        var formValue = request.formData[formKey]
+        if (formValue instanceof Array) {
```
[the remark this answers is not in the document]
Comment (codenirvana) [reply lib/redirect.js]: https://jsperf.com/isarray-vs-instanceof

File: tests/test-form-data-redirect.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+    req.on('data', function (d) {
+      data += d
+    })
+
+    req.on('end', function () {
+      // check for the fields' traces
+
+      if (options.batch) {
+        t.ok(data.indexOf('form-data; name="my_field_batch"') !== -1)
```
[the remark this answers is not in the document]
Comment (codenirvana) [reply tests/test-form-data-redirect.js]: Nope

File: lib/redirect.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (kunagpal) [reply lib/redirect.js]: Include this in the comment too 😄

File: lib/redirect.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (kunagpal) [reply lib/redirect.js]: 🐎

**Pull request merged**